File: staCMRx/staCMRx.m
function [x, f, stats] = staCMRx (data, model, E, shrink, tolerance, proc)
% function [x, f, stats] = staCMRx (data, model, E, shrink, proc)
% Multidimensional CMR
% data is cell array of data or structured output from staSTATS 
% model is a nvar*k matrix specifying the linear model; default = ones(nvar,1)
% E is partial order
% shrink is parameter to control shrinkage of covariance matrix;
% 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum
% returns:
% x = best fitting CMR values to y-means
% f = fit statistic
% shrinkage = estimated shrinkage of covariance matrix
% *************************************************************************
% previously modified 22 January 2016
% modified 12 April 2016 to circumvent null rows in model
% modified 3 June 2016 to deal with cell array data format
% modified 25 Aug 2016 to rectify a few odd things
% modified 12 Nov 2016 to include tolerance (0 to 1)
% *************************************************************************

if nargin < 6, proc = -1; end
if nargin < 5, tolerance=0; end
if nargin < 4, shrink=-1; end
if nargin < 3, E = {}; end
if nargin < 2, model=[]; end

if isempty(E), E = {}; end
if isempty(shrink), shrink=-1; end
if isempty(tolerance), tolerance=0; end
if isempty(proc), proc=-1; end

if ~iscell(E) && isvector(E)
    E = {E};
end

if iscell(data)
    if isstruct(data{1})
        y = data; % if structured then already in stats form
    else
        y = staSTATS(data, shrink); % otherwise assume within-subjects data and get stats
    end
else
    celldata = gen2cell (data); % convert from old "general format"
    y = staSTATS (celldata, shrink);
end

nvar = numel(y);
if isempty(model), model = ones(nvar,1); end % default model for STA

shrinkage = zeros(numel(y{1}.shrinkage),numel(y)); 
for ivar=1:nvar, shrinkage(:,ivar) = y{ivar}.shrinkage; end

% decompose model if there are null rows to short circuit result
x = cell(1,nvar);
s = any(model,2); i = find(s==1);

nitr=[]; adjStar=[]; fl=[]; fu=[];fuf=[]; remaining=[];
if rank(model(i,:))==numel(i)
    xx = cell(1,numel(i)); f = 0;
    for k=1:numel(i)
        xx{k} = y{i(k)}.means';
    end
else
    [xx, f, nitr, adjStar, fl, fu, fuf, remaining] = jCMRx (y(i), model(i,:), E, proc, tolerance); % call CMRx for non-zero rows of model
end

x(i) = xx;
j = find(s==0); % replace zero rows with weighted means
if ~isempty(j)
    for k=1:numel(j)
        u = j(k);
        w = y{u}.weights;
        if ~isvector(w)
            w = diag(w); if iscolumn(w), w=w'; end
        end
        m = y{u}.means; if isrow(m), m=m'; end
        z = w*m/sum(w);
        x{u} = z*ones(size(m));
        d = m - x{u};
        f = f + d'*diag(w)*d;
    end
end
if f < 0, f = 0; end

stats.shrinkage = shrinkage;
stats.nitr = nitr;
stats.adj = adjStar;
stats.fl = fl; 
stats.fu = fu;
stats.fuf = fuf;
stats.remaining = remaining;

    

File: STACMR-R/Utility functions/staCMRx.R
staCMRx <- function (data, model=NULL, E = list(), shrink=-1, tolerance=0, proc=-1, approx=0) {
  # function [x, f, shrinkage] = staCMRx (data, model, E, shrink, tolerance, proc, approx)
  # Multidimensional CMR
  # data is cell array of data or structured output from staSTATS 
  # model is a nvar*k matrix specifying the linear model; default = ones(nvar,1)
  # E is partial order
  # shrink is parameter to control shrinkage of covariance matrix;
  # 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum
  # tolerance ?
  # proc is number of processors (-1 means all possible)
  # approx ?
  # returns:
  # x = best fitting CMR values to y-means
  # fval = fit statistic
  # shrinkage = estimated shrinkage of covariance matrix
  
  # *************************************************************************
  # modified from matlab 14 September 2016
  # minor bug fixed 13 October 2016
  # updated 8 March 2017
  # *************************************************************************
  #
  # set up defaults
  tol <- 10e-5
  if (missing(E)) {E = list()}
  if (missing(shrink)) {shrink = -1}
  if (missing(tolerance)) {tolerance = 0}
  if (missing(proc)) {proc = -1}
  if (missing(approx)) {approx = 0}
  
  # get stats from data (depending on its form)
  if (is(data,"data.frame")) {
    y = gen2list (data) # convert from general format
    y = staSTATS (y, shrink) # get stats
  } else if (is.null(data[[1]]$means)) {y = staSTATS(data, shrink) # in list form, get stats
  } else {y = data} # already in stats form
  
  nvar = length(y)
  if (missing(model) || is.null(model)) {model = matrix(1,nvar,1)} # sta default model
  
  # extract shrinkage parameters (for information only)
  shrinkage = matrix(0, length(y[[1]]$shrinkage), nvar)
  for (ivar in 1:nvar) {shrinkage[,ivar] = y[[ivar]]$shrinkage}
  
  # decompose model if there are null rows to short circuit result
  s = rowSums(abs(model)); i = which(s != 0); nnonzero = length(i)
  xx = vector("list", nnonzero); fval = 0
  if (qr(model[i,])$rank==nnonzero){
    for (k in 1:nnonzero) {xx[[k]] = matrix(y[[i[k]]]$means,length(y[[i[k]]]$means),1)}
  } else {
    out = jCMRx (y[i], model[i,], E) # call jCMRx for non-zero rows of model
    fval = out$fval; 
    for (k in 1:nnonzero) {xx[[k]] = matrix(out$x[,k],nrow(out$x),1)}
  }
  
  # reassemble output
  x = vector("list", nvar); 
  for (k in 1:nnonzero) {x[[i[k]]] = xx[[k]]}
  
  # replace zero model rows with weighted means
  j = which(s == 0) 
  nzero = length(j)
  if (nzero > 0) {
    for (k in 1:nzero) {
      u = j[k]
      w = y[[u]]$weights
      if (is.matrix(w)) {w = diag(w)}
      w = matrix(w/sum(w),length(w),1)
      z = y[[u]]$means%*%w
      a = rep(z,length(y[[u]]$means))
      x[[u]] = matrix(a,length(y[[u]]$means),1)
      d = y[[u]]$means - t(x[[u]])
      fval = fval + d%*%y[[u]]$weights%*%t(d)
    }
  }
  
  if (fval < tol) {fval = 0} # round down
  
  output = list(x, fval, shrinkage)
  names(output) = c("x", "fval", "shrinkage")
  
  return(output)
}